### PR TITLE
BAH-3093 Refactor| Module Federation Micro-frontend Setup for IPD 

### PR DESCRIPTION
**Pull request by AshishkurianTw**

### CODE REVIEW

File: micro-frontends/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "openmrs-mfe-host",
+  "version": "0.0.1",
+  "description": "Hosting microfrontends for openmrs",
+  "main": "dist/index.js",
+  "author": "bahmni@thoughtworks.com",
+  "license": "AGPL-3.0",
```
Comment (rahu1ramesh) [inline micro-frontends/package.json:7]: @AshishkurianTw openmrs-module-bahmniapps is licensed under Apache-2.0. Should micro-frontends/ also adhere to the same license agreement, or is it acceptable to maintain AGPL-3.0 license?

cc: @angshu, @binduak

File: micro-frontends/package.json
Comment (rahu1ramesh) [inline micro-frontends/package.json:1]: openmrs-module-bahmniapps adheres to the specified engine compatibility standards as follows:
```
"engines": {
    "yarn": ">= 1.0.0",
    "npm": "use yarn instead of npm."
}
```
Is there a similar restriction on the micro-frontends code regarding engine compatibility?

File: micro-frontends/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "mini-css-extract-plugin": "^2.7.6",
+    "resolve-url-loader": "^3.1.0",
+    "webpack": "^5.86.0",
+    "webpack-cli": "^5.1.4",
+    "css-loader": "5.2.6",
+    "sass": "^1.63.4",
+    "sass-loader": "10.1.0"
+  },
+  "dependencies": {
+    "angular-component": "^0.1.3",
```
Comment (rahu1ramesh) [inline micro-frontends/package.json:24]: The latest published versions are:
- [bahmni-carbon-ui](https://www.npmjs.com/package/bahmni-carbon-ui): 0.1.9
- [carbon-components](https://www.npmjs.com/package/carbon-components): 10.58.12
- [carbon-components-react](https://www.npmjs.com/package/carbon-components-react): 8.54.0

If upgrading to the latest versions won't cause any significant issues or break the code, it would be great to use the latest version of all our dependencies.

File: micro-frontends/package.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    "carbon-components-react": "^7.25.0",
+    "react": "16.14.0",
+    "react-dom": "16.14.0"
+  },
+  "peerDependencies": {
+    "react-dom": "16.14.0",
+    "react2angular": "^4.0.6"
+  },
+  "scripts": {
+    "build": "webpack --mode production"
```
Comment (rahu1ramesh) [inline micro-frontends/package.json:37]: Can we also include a test coverage, test and lint script in the package.json file?

File: micro-frontends/src/ipd/index.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Adapt all IPD components to use within angular
+"use strict";
+
+import { react2angular } from "react2angular";
+import { IpdDashboard } from "./IpdDashboard";
+import { DrugChartModal } from "./DrugChartModal";
+import "bahmni-carbon-ui/styles.css";
```
Comment (rahu1ramesh) [inline micro-frontends/src/ipd/index.js:7]: @AshishkurianTw Could you kindly confirm if the addition of styles.css is relevant to this PR? It was not included in the previous PR, so I'm curious to understand if its inclusion was deliberate and the reasoning behind it.

File: micro-frontends/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ * An alternative to providing build time URLs
+ * We need to do this string promise stuff because we need to resolve the host at run-time.
+ * This is the way, as documented here: https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes
+ * 
+ * @param {string} name The name of the remote, as given in it's ModuleFederationPlugin configuration
+ * @param {string} subPath The sub-path of the remote, as set-up in the proxy configuration
+ * 
+ * @returns {string} A string that can be evaluated to a promise that resolves to the remote
+ */
+function dynamicRemote(name, subPath) {
```
Comment (rahu1ramesh) [inline micro-frontends/webpack.config.js:110]: Can you explain the rationale behind using `dynamicRemote` instead of `remoteProxiedAtHostDomain`?

File: micro-frontends/webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  resolve: {
+    extensions: [".tsx", ".ts", ".jsx", ".js", ".json"],
+    alias: {
+      react: path.resolve(__dirname, "./src/globalReact.js"),
+      'react-dom': path.resolve(__dirname, "./src/globalReactDom.js"),
+    }
+  },
+  entry: {
+    ipd: "./src/ipd/index.js",
+    mfe_polyfills_angular_1_4: "./src/polyfill.js",
```
Comment (rahu1ramesh) [inline micro-frontends/webpack.config.js:18]: @AshishkurianTw I remember in the earlier PR `"next-ui": "./src/next-ui/index.js",
    shared: "./src/shared.js"` was added to the entry instead of `mfe_polyfills_angular_1_4`. Just wanted to understand if this change was deliberate and the reasoning behind it.

File: ui/app/clinical/common/models/drugOrder.js
Comment (rahu1ramesh) [inline ui/app/clinical/common/models/drugOrder.js:1]: The changes added to this file here is for the IPD button feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/common/models/drugOrderViewModel.js
Comment (rahu1ramesh) [inline ui/app/clinical/common/models/drugOrderViewModel.js:1]: The changes added to this file here is for the IPD button feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/constants.js [hunk 7]
```diff
@@ -71,7 +75,8 @@ Bahmni.Clinical.Constants = (function () {
         globalPropertyToFetchActivePatients: 'emrapi.sqlSearch.activePatients',
         adtPrivilege: "app:adt",
         adtForwardUrl: "../adt/#/patient/{{patientUuid}}/visit/{{visitUuid}}/",
-        certificateHeader: "Print Header"
+        certificateHeader: "Print Header",
```
Comment (rahu1ramesh) [inline ui/app/clinical/constants.js:78]: Would it be possible to make `Print Header` a translation instead of keeping it hardcoded? This would enhance localization and maintain consistency across translations.

File: ui/app/clinical/constants.js [hunk 8]
```diff
@@ -18,6 +18,10 @@ Bahmni.Clinical.Constants = (function () {
         age: "Age",
         weight: "Weight"
     };
+    var careSetting = {
```
Comment (rahu1ramesh) [inline ui/app/clinical/constants.js:21]: The changes added to this file here is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/consultation/controllers/drugOrderHistoryController.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/controllers/drugOrderHistoryController.js:1]: The changes added to this file here is for the IPD Drug button feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/consultation/directives/newDrugOrders.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/directives/newDrugOrders.js:1]: The changes added to this file here is for the IPD Drug button feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/consultation/views/newDrugOrders.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/newDrugOrders.html:1]: The changes added to this file here is for the IPD Drug button feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html:1]: The changes added to this file here is for the `Print Medication` feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/displaycontrols/treatmentData/views/treatmentTable.html
Comment (rahu1ramesh) [inline ui/app/clinical/displaycontrols/treatmentData/views/treatmentTable.html:1]: The changes added to this file here is for the `IPD Drug button` feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/clinical/displaycontrols/treatmentData/views/treatmentTableRow.html
Comment (rahu1ramesh) [inline ui/app/clinical/displaycontrols/treatmentData/views/treatmentTableRow.html:1]: The changes added to this file here is for the `IPD Drug button` feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/i18n/clinical/locale_en.json
Comment (rahu1ramesh) [inline ui/app/i18n/clinical/locale_en.json:1]: This branch lag behind the master branch by a significant number of commits. It would greatly help the review process and mitigate potential merge conflicts if both branches were brought up to date with the master branch.

File: ui/app/ot/controller/listViewController.js
Comment (rahu1ramesh) [inline ui/app/ot/controller/listViewController.js:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

File: ui/app/ot/mappers/SurgicalBlockMapper.js
Comment (rahu1ramesh) [inline ui/app/ot/mappers/SurgicalBlockMapper.js:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

File: ui/app/ot/services/surgicalAppointmentService.js
Comment (rahu1ramesh) [inline ui/app/ot/services/surgicalAppointmentService.js:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

File: ui/app/ot/views/listView.html
Comment (rahu1ramesh) [inline ui/app/ot/views/listView.html:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

File: ui/app/styles/clinical/_patientDashboard.scss
Comment (rahu1ramesh) [inline ui/app/styles/clinical/_patientDashboard.scss:1]: The changes added to this file here is for the `IPD Drug button` feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/styles/clinical/treatment/_treatment.scss
Comment (rahu1ramesh) [inline ui/app/styles/clinical/treatment/_treatment.scss:1]: The changes added to this file here is for the `IPD Drug button` and `Print Medication` feature, which is not relevant to this PR. It would be appreciated if this could be a separate PR to allow us to track the changes more effectively, enabling both review and testing of this new feature.

File: ui/app/styles/ot/_ot.scss
Comment (rahu1ramesh) [inline ui/app/styles/ot/_ot.scss:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

File: ui/test/__mocks__/micro-frontends.js
Comment (rahu1ramesh) [inline ui/test/__mocks__/micro-frontends.js:1]: `angular.module("bahmni.mfe.nextUi", []);` seems to be missing in this PR. Just wanted to understand if this change was deliberate and the reasoning behind it.

File: ui/test/unit/ot/controller/listViewController.spec.js
Comment (rahu1ramesh) [inline ui/test/unit/ot/controller/listViewController.spec.js:1]: ***HEADS UP***

The changes added to this file here I believe is for the [BAH-3092](https://github.com/Bahmni/openmrs-module-bahmniapps/pull/903) feature, which is not relevant to this PR. The PR raised against BAH-3092 should first be reviewed and merged to master before we continue with this PR.

**Review by rahu1ramesh — CHANGES_REQUESTED** [review]

**Comment by rahu1ramesh** [comment]
@AshishkurianTw @tanyaa-tw I noticed that a few files related to the basic setup of the Module Federation Micro-frontend were omitted from the PR:

1. micro-frontends/.eslintrc.js
2. micro-frontends/jest.config.js
3. micro-frontends/src/next-ui/index.js
4. micro-frontends/src/next-ui/patientAlergies/PatientAlergiesControl.jsx
5. micro-frontends/src/setupTests.js
6. micro-frontends/src/shared.js
7. micro-frontends/src/styleMock.js
8. micro-frontends/src/utils/bridge-builder.js

I just wanted to understand if this omission was deliberate and if there was any specific reasoning behind it. Could you please provide some clarification on this matter?

**Comment by Phanindra-tw** [comment]
I've removed the MFE entry related to IPD as it is dependent on the IPD Container. The IPD container to be running successfully we would need quite a few changes in the Bahmni Apps. So once we have the dependency PRs merged, we will add back the IPD related bridge code.
cc: @rahu1ramesh @binduak @ayush34957

File: ui/app/clinical/dashboard/controllers/patientDashboardController.js [hunk 9]
```diff
@@ -13,9 +13,36 @@ angular.module('bahmni.clinical')
             $scope.visitSummary = visitSummary;
             $scope.enrollment = $stateParams.enrollment;
             $scope.isDashboardPrinting = false;
+            $scope.loadIPD = false;
             var programConfig = appService.getAppDescriptor().getConfigValue("program") || {};
             $state.discardChanges = false;
 
+            $scope.alergyData = {
```
Comment (rahu1ramesh) [inline ui/app/clinical/dashboard/controllers/patientDashboardController.js:20]: @AshishkurianTw @Phanindra-tw @tanyaa-tw I assume the changes made here will be revised later on. I believe this data was added for testing the micro-frontend setup. Please confirm, as everything else appears fine to me.

**Review by rahu1ramesh — APPROVED** [review]

File: ui/app/clinical/dashboard/controllers/patientDashboardController.js (hunk 9 above)
Comment (Phanindra-tw) [reply ui/app/clinical/dashboard/controllers/patientDashboardController.js:20]: Yes @rahu1ramesh. This is a temporary addition . The upcoming PRs will update these values.

**Review by rahu1ramesh — CHANGES_REQUESTED** [review]
It seems the commit (https://github.com/Bahmni/openmrs-module-bahmniapps/pull/907/commits/eb8bba3d4468454491455e16a6aeddd81e511067) in the micro-frontend PR may not be necessary and could cause issues during the merge process. Please review and remove the changes as appropriate.

**Comment by rahu1ramesh** [comment]
@Phanindra-tw @AshishkurianTw @tanyaa-tw From an earlier discussion, what I understood was that the changes related to drugOrder in the earlier PR seem to have been for testing the DrugChart on the microfrontend. However, the files are missing in the latest PR. Was this omission intentional? Please confirm.

Screenshot 2024-04-29 at 8 41 50 PM

File: ui/app/clinical/consultation/controllers/drugOrderHistoryController.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/controllers/drugOrderHistoryController.js:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

File: ui/app/clinical/consultation/directives/newDrugOrders.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/directives/newDrugOrders.js:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

File: ui/app/clinical/consultation/views/newDrugOrders.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/newDrugOrders.html:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

File: ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

File: ui/app/styles/clinical/treatment/_treatment.scss
Comment (rahu1ramesh) [inline ui/app/styles/clinical/treatment/_treatment.scss:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

File: ui/test/unit/clinical/controllers/drugOrderHistoryController.spec.js
Comment (rahu1ramesh) [inline ui/test/unit/clinical/controllers/drugOrderHistoryController.spec.js:1]: The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

**Review by rahu1ramesh — CHANGES_REQUESTED** [review]

File: ui/app/clinical/consultation/directives/newDrugOrders.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/directives/newDrugOrders.js:1]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

File: ui/app/clinical/consultation/views/newDrugOrders.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/newDrugOrders.html:1]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

File: ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/views/treatmentSections/drugOrderHistory.html:1]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

File: ui/app/clinical/dashboard/services/clinicalDashboardConfig.js
Comment (rahu1ramesh) [inline ui/app/clinical/dashboard/services/clinicalDashboardConfig.js:1]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

File: ui/app/styles/clinical/treatment/_treatment.scss [hunk 10]
```diff
@@ -169,6 +169,7 @@
             }
         }
 
+        button.ipd-btn,
```
Comment (rahu1ramesh) [inline ui/app/styles/clinical/treatment/_treatment.scss:172]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

File: ui/app/clinical/dashboard/views/dashboard.html
Comment (rahu1ramesh) [inline ui/app/clinical/dashboard/views/dashboard.html:1]: @Phanindra-tw is this change relevant for the mfe PR?

File: ui/app/clinical/consultation/controllers/drugOrderHistoryController.js
Comment (rahu1ramesh) [inline ui/app/clinical/consultation/controllers/drugOrderHistoryController.js:1]: @Phanindra-tw The IPD feature is still under discussion, and the changes added here to support it are not relevant to this PR. Please revert them.

cc: @binduak, @angshu

**Review by rahu1ramesh — CHANGES_REQUESTED** [review]

**Review by rahu1ramesh — APPROVED** [review]
